### PR TITLE
Implementation of Username Restrictions

### LINKED ISSUE #19563: Username bans
WizDen currently deals with bad usernames by placing an account ban, but since you can easily change usernames now this is not ideal. There are two primary purposes which should be served:

# Preemptive bans
In this use case, multiple blacklist regex are defined. Unlike notes and bans, these are likely to be modified frequently, at least in the time around their initial creation, so it may be beneficial to keep that in mind when designing a way to audit the changes. This would also benefit from a whitelist system, where specific usernames are provided that will be permitted even if they match a blacklist regex. Neither the regex blacklist nor the username whitelist strictly require the ability for admins to interact with them in-game, interaction through SS14.Admin is probably ideal.

# Reactive bans
In this use case, a player joins with a username that the game admins want to ban, so the specific player's username is blacklisted. This should be possible in-game.

# Ban message
The ban message for a username blacklist should probably be configurable, and each blacklist entry should probably have the ability to have a note added to it, especially regex blacklists, but a public per username ban reason is probably not needed. Players that hit a username ban should just be clearly told that their username isn't permitted and they can change it at (link) or appeal at (link).

### PR DESCRIPTION
**Pull request by benev0**
## About the PR

added a username rule table to the database
added a username whitelist table to the database
added a server ban manager with regex cache
added a client username ban cache manager for administrators 
added commands to manage bans `banusername` `banusernameregex` `pardonusername`
added commands to manage username whitelist `wladdusername` `wlrmusername`
added ui to interact with bans
altered connection manager to kick users on attempted join

[demo on yt](https://youtu.be/KAeCyBG3t5g)

## Why / Balance

resolves #19563
tl;dr: removes un-banning work for bans created for inappropriate usernames

## Technical details

- [x] create username rule in database schema
- [x] update sqllite and pgsql methods to include rule creation and removal
- [x] create pgsql username-rule triggers 
- [x] create rule manager service
- [x] create command (ban)
- [x] create command (pardon)
- [x] create command UI (ban)
- [x] create command UI (pardon) 
- [x] add command to admin UI
- [x] username allowlist

Service diagram (client manager has been moved into eui)
UsernameBan

## Breaking changes

**Changelog**

:cl:
ADMIN:
- add: Added username bans

### CODE REVIEW

**Comment by benev0** [comment]
I am gong to set the permission level to `AdminFlags.MassBan` for the command initially due to `[\s\S]*`.

### Feedback requested on my initial thoughts: 
-  simple username rule command (ban)
-  full regex username rule command (mass ban)
-  exempt admins from kick but log a violation
-  ensure that it is clear that this affects the server group and should not be used as a mass kick command

**Comment by benev0** [comment]
Sans pardon ui, username whitelists, and silly stuff like tests, core functionality is done. **Yay!**

The issue blocking pardon ui is that `MsgUsernameBans` when received by the client causes a disconnect. 
This is probably due to the monster list of 5-tuple (bool, bool, int, string, string) encoding and decoding.

**Comment by benev0** [comment]
these features are at minimum functional state as specified in #19563 
next required steps for merging of the PR as I understand them
- [ ] resolve merge conflict 
- [ ] revert and compress migrations into the minimum number of files
- [ ] implement required optimizations for non regex username bans
- [ ] implement any optional features 
  - [ ] extra full info ui in when viewing username bans
  - [ ] extend to ban functionality
- [ ] what level on server regex bans should be (currently mass ban) likely should be host
- [ ] items listed under Post Product (review)
  - [ ] manager tests
  - [ ] check localization support
  - [ ] check logging support
  - [ ] cleanup imports
  - [ ] check on server group

**Comment by benev0** [comment]
I was running a different build command locally I shall resolve this

File: Content.Server/Administration/Managers/UsernameRuleManager.cs [hunk 1]
```diff
@@ -0,0 +1,432 @@
...
+    [Dependency] private readonly IEntitySystemManager _systems = default!;
+
+    private ISawmill _sawmill = default!;
+
+    public const string SawmillId = "admin.username";
+
+    // this could be changed to sorted to implement deltas + removed bool or negative id
+    // id : (regex, regexString, message, ban)
+    private readonly Dictionary<int, (Regex?, string, string, bool)> _cachedUsernameRules = new();
+    private readonly HashSet<string> _cachedUsernames = new();
```
Comment (PJB3005) [inline Content.Server/Administration/Managers/UsernameRuleManager.cs:39]: This isn't used.

File: Content.Server/Administration/Managers/UsernameRuleManager.cs [hunk 2]
```diff
@@ -0,0 +1,432 @@
...
+        var usernameBanMsg = new MsgUsernameBan()
+        {
+            UsernameBan = new(id, false, false, false, "empty"),
+        };
+
+        _sawmill.Debug($"sent username ban delete {id} to active admins");
+        _net.ServerSendToMany(usernameBanMsg, _admin.ActiveAdmins.Select(a => a.Channel).ToList());
+    }
+
+    public void PostInject()
```
Comment (PJB3005) [inline Content.Server/Administration/Managers/UsernameRuleManager.cs:411]: This should be an explicit implementation.

File: Content.Server/Administration/Managers/UsernameRuleManager.cs [hunk 3]
```diff
@@ -0,0 +1,432 @@
...
+        _net.RegisterNetMessage<MsgFullUsernameBan>();
+        _net.RegisterNetMessage<MsgRequestFullUsernameBan>(OnRequestFullUsernameBan);
+
+        var rules = await _db.GetServerUsernameRulesAsync(false);
+
+        if (rules == null)
+        {
+            _sawmill.Warning("failed to get rules from database");
+            return;
+        }
```
Comment (PJB3005) [inline Content.Server/Administration/Managers/UsernameRuleManager.cs:60]: This check cannot fail according to nullable annotations.

File: Content.Server/Administration/Managers/UsernameRuleManager.cs [hunk 4]
```diff
@@ -0,0 +1,432 @@
...
+
+        foreach (var ruleDef in rules)
+        {
+            if (ruleDef.Id == null)
+            {
+                _sawmill.Warning("rule had Id of null");
+                continue;
+            }
```
Comment (PJB3005) [inline Content.Server/Administration/Managers/UsernameRuleManager.cs:68]: This should be an assert instead.

File: Content.Server.Database/Migrations/Postgres/20241017100853_UsernameBans.cs [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+
+                        perform pg_notify('username_rule_notification', json_build_object('username_rule_id', NEW.server_username_rule_id, 'server_id', x_server_id)::text);
+                        return NEW;
+                    end;
+                    $$ LANGUAGE plpgsql;
+            """);
+
+            migrationBuilder.Sql("""
+                create or replace trigger notify_on_server_ban_insert
+                    after update on server_username_rule
```
Comment (PJB3005) [inline Content.Server.Database/Migrations/Postgres/20241017100853_UsernameBans.cs:64]: After update? Shouldn't this be insert, delete and update?

File: Content.Server/Administration/Managers/UsernameRuleManager.Notification.cs [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+public sealed partial class UsernameRuleManager
+{
+    public const string UsernameRuleNotificationChannel = "username_rule_notification";
+
+    private static readonly TimeSpan UsernameRuleNotificationRateLimitTime = TimeSpan.FromSeconds(30);
+    private const int UsernameRuleNotificationRateLimitCount = 10;
+
+    private readonly object _usernameRuleNotificationRateLimitStateLock = new();
+    private TimeSpan _usernameRuleNotificationRateLimitStart;
+    private int _usernameRuleNotificationRateLimitCount;
```
Comment (PJB3005) [inline Content.Server/Administration/Managers/UsernameRuleManager.Notification.cs:16]: Daring approach, copy pasting this from the server ban code *without* the comments explaining why.

This rate limit doesn't make sense here, get rid of it.

File: Content.Server/Administration/Managers/UsernameRuleManager.Notification.cs [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+        }
+
+        if (!CheckRuleRateLimit())
+        {
+            _sawmill.Verbose("Not processing username rule notification due to rate limit");
+            return;
+        }
+
+        _taskManager.RunOnMainThread(() => ProcessUsernameRuleNotification(data));
+    }
```
Comment (PJB3005) [inline Content.Server/Administration/Managers/UsernameRuleManager.Notification.cs:50]: Unify this code with the one in BanManager you copy-pasted.

File: Content.Server/Administration/Managers/UsernameRuleManager.Notification.cs [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+            return;
+        }
+
+        if (usernameRule.Retired)
+        {
+            ClearCompiledRegex(usernameRule.Id ?? -1);
+            return;
+        }
+
+        CacheCompiledRegex(usernameRule.Id ?? -1, usernameRule.Regex, usernameRule.Expression, usernameRule.Message, usernameRule.ExtendToBan);
```
Comment (PJB3005) [inline Content.Server/Administration/Managers/UsernameRuleManager.Notification.cs:74]: Isn't this going to log a warning when updating an existing rule with a new regex? Is that a supported use case?

**Review by PJB3005 — CHANGES_REQUESTED** [review]
I didn't review any of the EUI/UI/client logic yet.

**Comment by benev0** [comment]
yeah, I have no clue perhaps it's a dependency tree issue? that is the only related change I can think of after reading the test.

**Comment by PJB3005** [comment]
It's likely just a heisentest i.e. completely unrelated to your changes, you can ignore it.

**Comment by benev0** [comment]
The client service still exists but the popup communication has been moved to the EUI.
The message should be formatted on connection denied but Cvars appears to not be acquiring the message correctly.
20241020_07h46m28s_grim

Otherwise I should be ready for additional feedback.

**Comment by benev0** [comment]
The current implementation uses `regexObj.Match(String)`, this checks for sub-string matches. If a full regex string match is needed then `"^MatchThisExactly$"` can be used. `regexObj.IsMatch(String)` could be used instead to fully match stings shifting the responsibility to match substrings to the rule author.

File: Content.Server/Administration/Managers/UsernameRuleManager.cs [hunk 9]
```diff
@@ -0,0 +1,429 @@
...
+
+        if (_cachedUsernameRules.ContainsKey(id))
+        {
+            _sawmill.Warning($"canceled caching attempt rule {id} already listed in cache");
+            return;
+        }
+
+        _sawmill.Info($"caching rule {id} {expression}");
+
+        var compiledRegex = regex ? new Regex(expression, RegexOptions.Compiled) : null;
```
Comment (PJB3005) [inline Content.Server/Administration/Managers/UsernameRuleManager.cs:137]: Don't make these regexes compiled.

**Review by PJB3005 — CHANGES_REQUESTED** [review]
There should still be no standalone cache manager on the client. Store it in the EUI state.

**Comment by benev0** [comment]
> There should still be no standalone cache manager on the client. Store it in the EUI state.

This is taking a lot longer than expected I am currently having issue with event subscription leakage on the client and server

**Comment by benev0** [comment]
@PJB3005 I will DM you after I check the PR diff, but otherwise, I have transferred the manager to the EUI dropping all custom net messages. I still have major concerns about the level of coupling in the new implementation.

**Comment by benev0** [comment]
looks like I need to split ccvars up

**Comment by benev0** [comment]
I cannot parse how these integration tests failed.

**Comment by Errant-4** [comment]
(Raised priority to match linked issue)

File: Content.Server/Administration/Systems/BwoinkSystem.cs [hunk 10]
```diff
@@ -180,6 +181,19 @@ private async void OnPlayerStatusChanged(object? sender, SessionStatusEventArgs
                     _activeConversations.Remove(e.Session.UserId);
                     return;
                 }
+
+                // check if the username has been banned
+                var lookup = await _playerLocator.LookupIdAsync(e.Session.UserId);
+
+                if (lookup != null) {
+                    var usernameBanReason = _usernameRules.IsUsernameBannedAsync(lookup.Username);
```
Comment (ElectroJr) [inline Content.Server/Administration/Systems/BwoinkSystem.cs:189]: `await`?  And for the `Loc.GetString`, shouldn't you pass `usernameBanReason.Message` instead of the whole struct, or does fluent somehow handle that.

File: Content.Server/Administration/UI/BanUsernamePanelEui.cs [hunk 11]
```diff
@@ -0,0 +1,183 @@
...
+        IoCManager.InjectDependencies(this);
+        _sawmill = _log.GetSawmill("admin.username_bans_eui");
+        _usernameRuleManager.UpdatedCache += SendSingleRule;
+    }
+
+    public override EuiStateBase GetNewState()
+    {
+        var hasBan = _admins.HasAdminFlag(Player, AdminFlags.Ban);
+        var hasHost = _admins.HasAdminFlag(Player, AdminFlags.Host);
+        return new BanUsernamePanelEuiState(hasBan, hasHost);
```
Comment (PJB3005) [inline Content.Server/Administration/UI/BanUsernamePanelEui.cs:29]: The client already knows its admin flags, you don't have to send these manually.

File: Content.Client/Administration/UI/UsernameBanInfoWindow.xaml [hunk 12]
```diff
@@ -0,0 +1,43 @@
...
+            <Label Name="CreateTimeLabel" Text="{Loc username-ban-panel-time-issued}" ClipText="True" HorizontalExpand="True"/>
+            <Label Name="CreateTimeContentLabel" Text="" ClipText="True" HorizontalExpand="True"/>
+
+            <Label Name="RoundIdLabel" Text="{Loc username-ban-panel-round}" ClipText="True" HorizontalExpand="True"/>
+            <Label Name="RoundIdContentLabel" Text="" ClipText="True" HorizontalExpand="True"/>
+
+            <Label Name="RetiredCheckboxLabel" Text="{Loc username-ban-panel-retired}" ClipText="True" HorizontalExpand="True"/>
+            <CheckBox Name="RetiredCheckbox" Access="Public" Disabled="True" Text="Retired" />
+
+            <Label Name="RetiringAdminLabel" Text="{Loc username-ban-panel-retirer}:" ClipText="True" HorizontalExpand="True"/>
```
Comment (PJB3005) [inline Content.Client/Administration/UI/UsernameBanInfoWindow.xaml:32]: Putting a `:` after the }` like this is not valid XAML.

File: Content.Client/Administration/UI/BanUsernamePanel/BanUsernamePanelEui.cs [hunk 13]
```diff
@@ -0,0 +1,117 @@
...
+        if (!update.Silent)
+        {
+            UpdatedCache?.Invoke(_usernameRulesCache.Values.ToList());
+        }
+    }
+
+    private void SubscribeList()
+    {
+        UpdatedCache += Window.ListController.PopulateList;
+        UpdatedCache.Invoke(_usernameRulesCache.Values.ToList());
```
Comment (PJB3005) [inline Content.Client/Administration/UI/BanUsernamePanel/BanUsernamePanelEui.cs:80]: This usage is extremely weird. Why not just make the call sites of `UpdatedCache` call `PopulateList` directly?

File: Content.Client/Administration/UI/BanUsernamePanel/BanUsernamePanelEui.cs [hunk 14]
```diff
@@ -0,0 +1,117 @@
...
+    }
+
+    public void RequestMoreInfo(int? ruleId)
+    {
+        if (ruleId is null)
+        {
+            return;
+        }
+
+        SendMessage(new BanUsernamePanelEuiMsg.GetRuleInfoRequest(ruleId ?? -1));
```
Comment (PJB3005) [inline Content.Client/Administration/UI/BanUsernamePanel/BanUsernamePanelEui.cs:101]: Why can this ever use the ID `-1`?

Comment (PJB3005) [inline Content.Client/Administration/UI/BanUsernamePanel/BanUsernamePanelEui.cs:101]: Why can this ever use the ID `-1`?

**Comment by PJB3005** [comment]
GitHub did the "send review early" thing again. Stand by.

File: Content.Server/Administration/Managers/UsernameRuleManager.cs [hunk 15]
```diff
@@ -0,0 +1,305 @@
...
+    }
+
+    private void KickMatchingConnectedPlayers(int id, ServerUsernameRuleDef def, string source)
+    {
+        if (!_cachedUsernameRules.ContainsKey(id))
+        {
+            return;
+        }
+
+        (Regex? compiledRule, string expression, _, _) = _cachedUsernameRules[id];
```
Comment (PJB3005) [inline Content.Server/Administration/Managers/UsernameRuleManager.cs:256]: This code does not feel like it's made easier to read by doing this deconstruction.

File: Content.Server/Database/ServerUsernameRuleDef.cs [hunk 16]
```diff
@@ -0,0 +1,84 @@
+using Content.Shared.CCVar;
+using Robust.Shared.Configuration;
+using Robust.Shared.Network;
+
+namespace Content.Server.Database;
+
+/// <summary>
+/// This class is used as I/O for the database manager
+/// </summary>
```
Comment (PJB3005) [inline Content.Server/Database/ServerUsernameRuleDef.cs:9]:
```suggestion
/// <summary>
/// Represents a copy of a username ban record in the database.
/// </summary>
```

File: Content.Server/Database/ServerUsernameRuleDef.cs [hunk 17]
```diff
@@ -0,0 +1,84 @@
...
+    public static string FormatUsernameViolationMessage(IConfigurationManager cfg, ILocalizationManager loc, string message)
+    {
+        var change = cfg.GetCVar(CCVars.InfoLinksChangeUsername);
+        return $"""
+                {loc.GetString("restrict-username-1")}
+                {loc.GetString("restrict-username-2", ("reason", message))}
+                {loc.GetString("restrict-username-3")}
+                {loc.GetString("restrict-username-4", ("link", change))}
+                """;
+    }
```
Comment (PJB3005) [inline Content.Server/Database/ServerUsernameRuleDef.cs:83]: It's not clear to me why both of these need to exist. The one that takes a `string message` doesn't ever get called without just manually passing `Message` manually.

File: Content.Server/Database/ServerUsernameRuleDef.cs [hunk 18]
```diff
@@ -0,0 +1,84 @@
...
+    /// <param name="loc"></param>
+    /// <returns>the formatted username ban message</returns>
+    public string FormatUsernameViolationMessage(IConfigurationManager cfg, ILocalizationManager loc)
+    {
+        return $"""
+                {loc.GetString("restrict-username-1")}
+                {loc.GetString("restrict-username-2", ("reason", Message))}
+                {loc.GetString("restrict-username-3")}
+                """;
+    }
```
Comment (PJB3005) [inline Content.Server/Database/ServerUsernameRuleDef.cs:64]: You're not formatting the link here.

File: Content.Server/EUI/BaseEui.cs [hunk 19]
```diff
@@ -92,7 +92,7 @@ public void SendMessage(EuiMessageBase message)
         /// <summary>
         ///     Close the EUI, breaking the connection between client and server.
         /// </summary>
-        public void Close()
+        public virtual void Close()
```
Comment (PJB3005) [inline Content.Server/EUI/BaseEui.cs:95]: Why can't you just... override `Closed()`?

File: Resources/Locale/en-US/connection-messages.ftl [hunk 20]
```diff
@@ -22,6 +22,11 @@ ban-banned-1 = You, or another user of this computer or connection, are banned f
 ban-banned-2 = The ban reason is: "{$reason}"
 ban-banned-3 = Attempts to circumvent this ban such as creating a new account will be logged.
 
+restrict-username-1 = Your username is prohibited either specifically or by class.
+restrict-username-2 = The reason is: {$reason}
+restrict-username-3 = You may be welcome back on the server after your name is changed.
```
Comment (PJB3005) [inline Resources/Locale/en-US/connection-messages.ftl:27]: Isn't this not always correct given that we have "extend to ban" bans here?

File: Resources/Locale/en-US/info/ban.ftl [hunk 21]
```diff
@@ -85,3 +85,61 @@ server-ban-string-never = never
...
+username-ban-panel-retired = Retired:
+username-ban-panel-retirer = Retiring admin:
+username-ban-panel-tine-retired = Retired Time:
+username-ban-panel-message = Message:
+
+cmd-username-whitelist-not-in-db = this username was not present in the database
+cmd-whitelist-username-help = Usage: {$command} <username>
+
+cmd-pardonusername-help = Usage: pardonusername <ban id 1> <ban id 2> ...
+cmd-pardonusername-send = Sending retire request rule {$id}
```
Comment (PJB3005) [inline Resources/Locale/en-US/info/ban.ftl:119]: This is very weird wording, why not just say what happened?

File: Content.Server/Administration/UI/BanUsernamePanelEui.cs [hunk 22]
```diff
@@ -0,0 +1,183 @@
...
+    [Dependency] private readonly ILogManager _log = default!;
+    [Dependency] private readonly IAdminManager _admins = default!;
+
+    private readonly ISawmill _sawmill;
+
+    public BanUsernamePanelEui()
+    {
+        IoCManager.InjectDependencies(this);
+        _sawmill = _log.GetSawmill("admin.username_bans_eui");
+        _usernameRuleManager.UpdatedCache += SendSingleRule;
```
Comment (PJB3005) [inline Content.Server/Administration/UI/BanUsernamePanelEui.cs:22]: You should subscribe to this in `Opened()`.

File: Content.Server/Administration/UI/BanUsernamePanelEui.cs [hunk 23]
```diff
@@ -0,0 +1,183 @@
...
+            return;
+        }
+
+        for (int i = 0; i < data.Count - 1; i++)
+        {
+            SendSingleRule(data[i], true);
+        }
+
+        var ultimate = data[data.Count - 1];
+        SendSingleRule(ultimate, false);
```
Comment (PJB3005) [inline Content.Server/Administration/UI/BanUsernamePanelEui.cs:72]: Make this a single net message that contains all the rules at once.

File: Content.Server/Administration/UI/BanUsernamePanelEui.cs [hunk 24]
```diff
@@ -0,0 +1,183 @@
...
+    private async void SendFullUsernameInfo(int ruleId)
+    {
+        var banData = await _usernameRuleManager.GetFullBanInfoAsync(ruleId);
+        if (banData is null)
+        {
+            return;
+        }
+        var message = new BanUsernamePanelEuiMsg.FullUsernameRuleInfoReply(
+            banData.CreationTime.UtcDateTime,
+            banData.Id ?? -1,
```
Comment (PJB3005) [inline Content.Server/Administration/UI/BanUsernamePanelEui.cs:163]: When can this ever be null?

**Review by PJB3005 — CHANGES_REQUESTED** [review]

**Comment by Simyon264** [comment]
o7

File: Resources/Locale/en-US/connection-messages.ftl (hunk 20 above)
Comment (benev0) [reply Resources/Locale/en-US/connection-messages.ftl:27]: bans are checked before username restrictions in the login manager, if a user is banned then the user will not see this message.

Comment (PJB3005) [reply Resources/Locale/en-US/connection-messages.ftl:27]: Will the ban message take priority if the user gets on-the-spot banned in this case?

**Comment by TheShuEd** [comment]
This PR has been in “Awaiting changes” status for over 3 months now without any visible updates, so I'm closing it as derelict. If you nevertheless plan to continue, please comment it and the PR will be reopened.

If you need help, you can ask questions in discord or in this PR

**Comment by benev0** [comment]
Hey, thanks for the reviews, @PJB3005; however, this should remain closed but can be picked up somewhere else. For the sake of thoroughness I will state what is wrong with this PR.

### Trivial flaws
These are issues stemming from my bad practices (see last review) 

1. Code Quality
Many of the problems in the code summited stem from the original code which was replicated and haphazardly morphed into a solution. As such there are issues from both the original code and new issues introduced in the alterations. 

2. Dead Code
The username ban to full ban feature was abandoned and never fully removed.

### Fundamental flaws
These are issues stemming from the design choices I made. 

This feature is banning wants to be banning, but it is implemented as a separate feature. This has resulted in the addition of commands nearly identical to the banning commands, UI separate from the banning UI, and UsernameBanManagers separate from BanManagers. 

1. Separation of Banning
Banning needs to be unified, certainty in UI, for obvious admin reasons. Moreover the goal of multi-target banning, that is the banning of a user's account, IP, username, and hardware identification, conflicts with a separation of feature complete banning systems.

2. Server Resources
i. Caching all regex username bans on all servers gets messy as users get creative.
ii. Executing arbitrary regex is unsafe

3. Regex is Overkill
Simple substring matches will correctly catch most cases. In the event that substrings are over zealous, an bypass allowlist can be added. Allowlists where also required for regex. 

For reasons 2 & 3 Regex should be nixed; this is a deviation from the specification in #19563.
